src/transformers/models/zamba/modeling_zamba.py

```diff
@@ -0,0 +1,1739 @@
+                        attention_mask=causal_mask,
+                        position_ids=position_ids,
+                        past_key_value=past_key_values,
+                        output_attentions=output_attentions,
+                        use_cache=use_cache,
+                        cache_position=cache_position,
+                    )
+                transformer_hidden_states = layer_outputs[0]
+                if output_attentions:
+                    if layer_outputs[1] is not None:
+                        all_self_attns += (layer_outputs[1],)
+                if self.gradient_checkpointing and self.training:
+                    transformer_hidden_states = self._gradient_checkpointing_func(
+                        next(linear_layers).__call__,
+                        transformer_hidden_states,
+                    )
+                else:
+                    transformer_hidden_states = next(linear_layers)(
+                        transformer_hidden_states,
+                    )
+            else:
+                transformer_hidden_states = None
+
+            if self.gradient_checkpointing and self.training:
+                layer_outputs = self._gradient_checkpointing_func(
+                    next(mamba_layers).__call__,
+                    hidden_states,
+                    transformer_hidden_states,
+                    causal_mask,
+                    position_ids,
+                    past_key_values,
+                    output_attentions,
+                    use_cache,
+                    cache_position,
+                )
+            else:
+                layer_outputs = next(mamba_layers)(
+                    hidden_states,
+                    transformer_hidden_states=transformer_hidden_states,
+                    attention_mask=causal_mask,
+                    position_ids=position_ids,
+                    past_key_value=past_key_values,
+                    output_attentions=output_attentions,
+                    use_cache=use_cache,
+                    cache_position=cache_position,
+                )
+            hidden_states = layer_outputs[0]
+
+        hidden_states = self.final_layernorm(hidden_states)
+
+        # add hidden states from the last decoder layer
+        if output_hidden_states:
+            all_hidden_states += (hidden_states,)
+
+        if past_key_values and not past_key_values.has_previous_state:
+            past_key_values.has_previous_state = True
+
+        next_cache = None if not use_cache else past_key_values
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, next_cache, all_hidden_states, all_self_attns] if v is not None)
+
+        return BaseModelOutputWithPast(
+            last_hidden_state=hidden_states,
+            past_key_values=next_cache,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attns,
+        )
+
+    def _update_causal_mask(self, attention_mask, input_tensor, cache_position):
```

Comment:
This can be copied from jamba: https://github.com/huggingface/transformers/blob/0ed3ffcb4461a244b87781a24e5ebd0a78f98142/src/transformers/models/jamba/modeling_jamba.py#L1518

Reply:
I added the comment `# Copied from transformers.models.jamba.modeling_jamba.JambaModel._update_causal_mask` in the line above https://github.com/Zyphra/transformers_zamba/blob/main/src/transformers/models/zamba/modeling_zamba.py#L1384C4-L1384C90